lib/services/version-service.ts

```diff
@@ -2,41 +2,42 @@ import { join } from "path";
 import * as semver from "semver";
 
 export class VersionService implements IVersionService {
-	private static DEFAULT_CLI_VERSION = "3.0.0";
-
 	constructor(private $fs: IFileSystem,
 		private $httpClient: Server.IHttpClient,
-		private $logger: ILogger) { }
+		private $logger: ILogger,
+		private $projectDataService: IProjectDataService) { }
 
 	public async getCliVersion(runtimeVersion: string): Promise<string> {
 		try {
 			const latestMatchingVersion = process.env.TNS_CLI_CLOUD_VERSION || await this.getLatestMatchingVersion("nativescript", this.getVersionRangeWithTilde(runtimeVersion));
-			return latestMatchingVersion || VersionService.DEFAULT_CLI_VERSION;
+			if (!latestMatchingVersion) {
+				throw new Error(`Cannot find matching CLI version based on runtime version ${runtimeVersion}`);
+			}
+
+			return latestMatchingVersion;
 		} catch (err) {
 			this.$logger.trace(`Unable to get information about CLI versions. Error is: ${err.message}`);
+			if (!semver.valid(runtimeVersion)) {
+				// In case the runtime version is not a valid semver version, we cannot determine CLI version.
+				throw new Error("Unable to determine CLI version for cloud build.");
+			}
+
 			return `${semver.major(runtimeVersion)}.${semver.minor(runtimeVersion)}.0`;
 		}
 	}
 
-	public async getRuntimeVersion(platform: string, nativescriptData: any, coreModulesVersion: string): Promise<string> {
+	public async getProjectRuntimeVersion(projectDir: string, platform: string, ): Promise<string> {
 		const runtimePackageName = `tns-${platform.toLowerCase()}`;
-		let runtimeVersion = nativescriptData && nativescriptData[runtimePackageName] && nativescriptData[runtimePackageName].version;
-		if (!runtimeVersion && coreModulesVersion) {
-			// no runtime added. Let's find out which one we need based on the tns-core-modules.
-			if (semver.valid(coreModulesVersion)) {
-				runtimeVersion = await this.getLatestMatchingVersion(runtimePackageName, this.getVersionRangeWithTilde(coreModulesVersion));
-			} else if (semver.validRange(coreModulesVersion)) {
-				// In case tns-core-modules in package.json are referred as `~x.x.x` - this is not a valid version, but is valid range.
-				runtimeVersion = await this.getLatestMatchingVersion(runtimePackageName, coreModulesVersion);
-			}
+		const runtimeVersion = this.$projectDataService.getNSValue(projectDir, `${runtimePackageName}.version`);
+
+		if (!runtimeVersion) {
+			throw new Error(`Unable to find runtime version for package ${runtimePackageName}.`);
 		}
 
-		return runtimeVersion || VersionService.DEFAULT_CLI_VERSION;
+		return runtimeVersion;
 	}
 
 	public getCoreModulesVersion(projectDir: string): string {
```

Comment:
Let's delete this method as it currently remains unused.
My vote goes towards removing redundant code and introducing it subsequently should the need arise